### PR TITLE
Create backend server and connect to recipes API

### PR DESCRIPTION
**Pull request by arisachue**
- Create backend server using Node.js
- Pass in user's current products as params to `"http://localhost:3001/apirecipes/"` e.g. `"http://localhost:3001/apirecipes/apple,+orange,+grape"`
- Connect to API from backend and fetch recipes based on the params. Store the recipes at endpoint
- Get the generated recipes from the frontend and pass it into recipes component to display the recipes
- Recipes will generate every time user changes their products

When the pantry has apple, orange, grape, bacon, milk, pasta:
Screen Shot 2022-07-08 at 11 39 27 AM

When user removes pasta and bacon:
Screen Shot 2022-07-08 at 11 40 47 AM

### CODE REVIEW

File: src/components/App/App.jsx [hunk 1]
```diff
@@ -56,6 +60,30 @@ const Operations = Object.freeze({
...
+    try {
+      // API parameter format: ingredient,+ingredient,+ingredient
+      let ingredientParams = ""
+      for (let i = 0; i < products.length; i++) {
+        if (i == 0) {
+          ingredientParams += products[i].name
+        } else {
+          ingredientParams += `,+${products[i].name}`
+        }
+      }
```
Comment (mjquigley) [inline src/components/App/App.jsx]: You should be able to accomplish this with something like this:
```
products.join()
```
That should take an array of items and join it with a separator. The default separator is `,`
If you just need the product name you may have to do a `.map()` first to map product to product.name and then join.

File: src/components/App/App.jsx [hunk 2]
```diff
@@ -42,12 +43,15 @@ export default function App() {
 
   const sampleRecipeInfo = constRecipeInfo
 
+  const api_url = "http://localhost:3001/apirecipes/"
```
Comment (mjquigley) [inline src/components/App/App.jsx]: nit: maybe `recipeApiUrl` to make this more clear when reading the code that uses it. Especially since you may add other APIs in the future.

File: src/components/ProductCard/ProductCard.jsx [hunk 3]
```diff
@@ -2,19 +2,22 @@ import * as React from "react"
...
-
+  /*
+  operations enum format:
+    const operations = Object.freeze({
+      Add: Symbol("add"),
+      Subtract: Symbol("subtract"),
+      Erase: Symbol("erase")
+    })
+  */
+
```
Comment (mjquigley) [inline src/components/ProductCard/ProductCard.jsx]: You can delete this if you don't need it. You can always look at the code history in git/github if you need to find old code.

File: src/components/App/App.jsx [hunk 4]
```diff
@@ -146,8 +177,8 @@ const handleGetRecipeInstructions = (recipeId) => {
           <Routes>
             <Route path="/" element={<Home />}/>
             <Route path="/profile" element={<Profile />}/>
-            <Route path="/recipes" element={<Recipes recipeinfo={recipeinfo} handleRecipeCardClick={handleRecipeCardClick} recipeInstructions={recipeInstructions} handleRecipeModal={handleRecipeModal} recipeModelShow={recipeModelShow} recipes={sampleRecipes}/>}/>
-            <Route path="/pantry" element={<Pantry productForm={productForm} handleOnProductFormChange={handleOnProductFormChange} handleOnSubmitProductForm={handleOnSubmitProductForm} handleProductQuantity={handleProductQuantity} products={products}/>}/>
+            <Route path="/recipes" element={<Recipes recipeinfo={recipeinfo} handleRecipeCardClick={handleRecipeCardClick} recipeInstructions={recipeInstructions} handleRecipeModal={handleRecipeModal} recipeModelShow={recipeModelShow} recipes={recipes}/>}/>
+            <Route path="/pantry" element={<Pantry operations={Operations} productForm={productForm} handleOnProductFormChange={handleOnProductFormChange} handleOnSubmitProductForm={handleOnSubmitProductForm} handleProductQuantity={handleProductQuantity} products={products}/>}/>
```
Comment (mjquigley) [inline src/components/App/App.jsx:171]: Just curious, what motivated moving Operations into this property?

File: express-server/app.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+const app = express()
+
+const api_key = "a881bae80f664370858fb9bbfc57a42d"
+
+app.use(express.json())
+app.use(morgan("tiny"))
+app.use(cors())
+
+app.get('/apirecipes/:ingredients', async (request, response) => {
+    console.log(request.params)
```
Comment (reddysindhu1) [inline express-server/app.js]: Is there a reason these console logs are here, do you need them further debugging?

File: src/components/App/App.jsx [hunk 6]
```diff
@@ -42,12 +43,15 @@ export default function App() {
...
+
 // state variables
 const [products, setProducts] = useState(basicProducts)
 const [productForm, setProductForm] = useState(basicProductForm)
-const [recipeModelShow, setRecipeModalShow] = useState(false);
-const [recipeInstructions, setRecipeInstructions] = useState([]);
-const [recipeinfo, setRecipeinfo] = useState({});
+const [recipeModelShow, setRecipeModalShow] = useState(false)
+const [recipeInstructions, setRecipeInstructions] = useState([])
+const [recipeinfo, setRecipeinfo] = useState({})
```
Comment (reddysindhu1) [inline src/components/App/App.jsx]: Nit: recipeinfo and setRecipeinfo isn't camelCase.

File: src/components/App/App.jsx [hunk 7]
```diff
@@ -42,12 +43,15 @@ export default function App() {
```
Comment (reddysindhu1) [inline src/components/App/App.jsx:41]: Github is not letting me comment on line 40 but since you're no longer using sampleRecipes, you can remove it.

File: src/components/App/App.jsx [hunk 8]
```diff
@@ -56,6 +60,30 @@ const Operations = Object.freeze({
...
+      // API parameter format: ingredient,+ingredient,+ingredient
+      let ingredientParams = ""
+      for (let i = 0; i < products.length; i++) {
+        if (i == 0) {
+          ingredientParams += products[i].name
+        } else {
+          ingredientParams += `,+${products[i].name}`
+        }
+      }
+      const api_ingredients = api_url + ingredientParams
```
Comment (reddysindhu1) [inline src/components/App/App.jsx]: Nit: The casing here isn't consistent with your other variables, make this maybe recipeApiIngredients

File: src/components/App/App.jsx [hunk 9]
```diff
@@ -64,21 +92,24 @@ const handleProductQuantity = (productName, operation) => {
 
   // edit item quantity
   //TODO: throw error if item didn't exist in products
-  if (operation === Operations.Add) {
```
Comment (reddysindhu1) [inline src/components/App/App.jsx]: Why did you change this from using strict equality? You don't have to change it I'm just curious.

File: src/components/App/App.jsx (hunk 1 above)
Comment (arisachue) [reply src/components/App/App.jsx]: Thanks for the suggestion, this works! I changed it to
`let ingredientParams = products.map((product) => (product.name)).join(",+")`

File: src/components/App/App.jsx (hunk 4 above)
Comment (arisachue) [reply src/components/App/App.jsx:171]: Since I was creating each `"add"` `"subtract"` `"erase"` strings into a `Symbol`, I was creating a unique value for each string. When I had two separate Operations enums (one in App.jsx and in ProductCard.jsx), even when the operations were the same, I was comparing two different values. I had to pass in the same Operations enum to make sure I was comparing the same values

File: express-server/app.js (hunk 5 above)
Comment (arisachue) [reply express-server/app.js]: Nope, I will take those out

File: src/components/App/App.jsx (hunk 6 above)
Comment (arisachue) [reply src/components/App/App.jsx]: I initially had it as `recipeInfo` and `setRecipeInfo` but the console.log was showing warnings that `recipeInfo` suggests the variable to be a DOM element, so I changed the capitalization.
Screen Shot 2022-07-08 at 2 53 15 PM
I admit that was a lazy fix so I researched what the error meant, and it seems that it's because I pass in all the `props` in RecipeModal.jsx in the `Modal` component, and `recipeInfo` is not a valid attribute to a `Modal` component. I found a fix, which is including this line:
`const {recipeInfo, instructions, title, ...modalProps} = props`
and pass in `{...modalProps}` in the `Modal` component

File: src/components/App/App.jsx (hunk 9 above)
Comment (arisachue) [reply src/components/App/App.jsx]: I was trying to figure out why my operations were not working, and I tried to see if the strict equality was the reason. Turns out it was because the `Symbol` function (I mentioned the fix in 5 comments above). I'll change it back since I think strict equality is better code

**Review by reddysindhu1 — APPROVED** [review]
Changes look good!
